src/v/kafka/server/usage_manager.cc

```diff
@@ -0,0 +1,386 @@
+      "Error correctly detecting last window delta");
+    _timer.arm((now + _usage_window_width_interval) - last_window_delta);
+    _persist_disk_timer.arm(now + _usage_disk_persistance_interval);
+}
+
+std::vector<usage_window>
+usage_manager::accounting_fiber::get_usage_stats() const {
+    std::vector<usage_window> stats;
+    for (size_t i = 1; i < _buckets.size(); ++i) {
+        const auto idx = (_current_window + i) % _usage_num_windows;
+        if (!_buckets.at(idx).is_uninitialized()) {
+            stats.push_back(_buckets.at(idx));
+        }
+    }
+    /// Open bucket last ensures ordering from oldest to newest
+    stats.push_back(_buckets.at(_current_window));
+    /// std::reverse returns results in ordering from newest to oldest
+    std::reverse(stats.begin(), stats.end());
+    return stats;
+}
+
+ss::future<> usage_manager::accounting_fiber::stop() {
+    _timer.cancel();
+    _persist_disk_timer.cancel();
+    co_await _gate.close();
+    try {
+        co_await persist_to_disk(
+          _kvstore,
+          persisted_state{
+            .configured_period = _usage_window_width_interval,
+            .configured_windows = _usage_num_windows,
+            .current_state = _buckets});
+    } catch (const std::exception& ex) {
+        vlog(
+          klog.debug,
+          "Encountered exception when persisting usage data to disk: {}",
+          ex);
+    }
+}
+
+ss::future<> usage_manager::accounting_fiber::close_window() {
+    const auto now = ss::lowres_system_clock::now();
+    _buckets.at(_current_window).end = epoch_time_secs(now);
+    _buckets.at(_current_window).u = co_await _um.map_reduce0(
+      [](usage_manager& um) { return um.sample(); },
+      _buckets.at(_current_window).u,
+      [](const usage& acc, const usage& x) { return acc + x; });
+    _current_window = (_current_window + 1) % _buckets.size();
+    _buckets.at(_current_window).reset(now);
+}
+
+std::chrono::seconds usage_manager::accounting_fiber::reset_state(
+  std::vector<usage_window> buckets) {
+    /// called after restart to determine which bucket is the 'current' bucket
+    auto last_window_delta = std::chrono::seconds(0);
+    _current_window = 0;
+    if (!buckets.empty()) {
+        std::optional<size_t> open_index;
+        for (size_t i = 0; i < buckets.size(); ++i) {
+            /// There will always be exactly 1 open_window in the result set
+            if (buckets.at(i).is_open()) {
+                vassert(!open_index, "Data serialization was incorrect");
+                open_index = i;
+                break;
+            }
+        }
+        vassert(open_index, "Data serialization was incorrect");
+        /// Optimization to begin picking up if wall time is within
+        /// window interval
+        const auto begin = std::chrono::seconds(buckets.at(*open_index).begin);
+        const auto now = std::chrono::seconds(epoch_time_secs());
+        vassert(begin < now, "Clock or data serialization error");
```

Comment:
two reasons to think about this assertion carefully:
* could they be equal? with low-res clocks, a lot can happen within the resolution of the clock ticks
* you're using a system clock, which ostensibly means that the clock can move forward and backward.